src/codegen/statements.rs

```diff
@@ -63,21 +63,84 @@ pub(crate) fn statement(
                     vartab,
                     opt,
                 };
-
                 //If we remove the assignment, we must keep expressions that have side effects
                 init.recurse(&mut params, process_side_effects_expressions);
                 return;
             }
 
             let expr = expression(init, cfg, contract_no, Some(func), ns, vartab, opt);
+
+            let location = *loc;
+            let pos = *pos;
+
             cfg.add(
                 vartab,
                 Instr::Set {
-                    loc: *loc,
-                    res: *pos,
-                    expr,
+                    loc: location,
+                    res: pos,
+                    expr: expr.clone(),
                 },
             );
+
+            //Lets check if the declaration is a declaration of a dynamic array
+            if let Expression::AllocDynamicArray(_loc, _ty, size, _option) = expr {
+                //get the size expression: numberLiteral/variable/Cast
+                let size_unboxed = *size;
+
+                //if its a variable of uint32, assign it to a temp
+                if let Expression::Variable(_loc, ty, position) = size_unboxed.clone() {
+                    let num = Expression::Variable(pt::Loc::Codegen, Type::Uint(32), position);
+                    let temp_res = vartab.temp_name("array_size", &ty);
+                    cfg.add(
+                        vartab,
+                        Instr::Set {
+                            loc: location,
+                            res: temp_res,
+                            expr: num.clone(),
+                        },
+                    );
+
+                    cfg.array_lengths_temps.insert(pos, (temp_res, num));
+                }
```

Comment:
This could be a chain of `if-elses` or a `match` statement, because all the conditions are exclusive.